library/Fission/Storage/IPFS/Add.hs

```diff
@@ -0,0 +1,178 @@
+        [cid] ->
+          cid
+            |> UTF8.textShow
+            |> UTF8.stripN 1
+            |> mkCID
+            |> IPFS.Pin.add
+
+        bad ->
+          return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+addFile ::
+  ( MonadRIO          cfg m
+  , HasProcessContext cfg
+  , HasLogFunc        cfg
+  , Has HTTP.Manager  cfg
+  , Has IPFS.URL      cfg
+  , Has IPFS.BinPath  cfg
+  , Has IPFS.Timeout  cfg
+  )
+  => Lazy.ByteString
+  -> IPFS.Name
+  -> m (Either IPFS.Error.Add IPFS.SparseTree)
+addFile raw name =
+  IPFS.Proc.run opts raw >>= \case
+    (ExitSuccess, result, _) ->
+      IPFS.Pin.add (mkCID <| UTF8.textShow result) >>= \case
+        Left err ->
+          return . Left . UnknownAddErr <| UTF8.textShow err
+
+        Right _ ->
+          case CL.lines result of
+            [inner, outer] ->
+              let
+                sparseTree  = Directory [(Hash rootCID, fileWrapper)]
+                fileWrapper = Directory [(fileName, Content fileCID)]
+                rootCID     = CID <| UTF8.textShow outer
+                fileCID     = CID . UTF8.stripN 1 <| UTF8.textShow inner
+                fileName    = Key name
+              in
+                return <| Right sparseTree
+
+            bad ->
+              return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+    where
+      opts = [ "add"
+             , "-wq"
+             , "--stdin-name"
+             , unName name
+             ]
+
+addPath ::
+  ( RIOProc           cfg m
+  , Has IPFS.Timeout  cfg
+  , Has IPFS.BinPath  cfg
+  )
+  => FilePath
+  -> m (Either IPFS.Error.Add CID)
+addPath path = IPFS.Proc.run ["add", "-HQ", path] "" >>= pure . \case
+    (ExitSuccess, result, _) ->
+      case CL.lines result of
+        [cid] -> Right . mkCID . UTF8.stripN 1 <| UTF8.textShow cid
+        bad   -> Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      Left . UnknownAddErr <| UTF8.textShow err
+
+addDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => IPFS.Ignored
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+addDir ignored path = doesFileExist path >>= \case
+  True -> addPath path
+  False -> walkDir ignored path
+
+walkDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => IPFS.Ignored
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+walkDir ignored path = do
+  files <- listDirectory path
+  let toAdd = removeIgnored ignored files
+  let f = foldResults path ignored
+  foldM f (Right <| Node { dataBlock = "CAE=", links = [] }) toAdd >>= \case
+    Left err -> return <| Left err
+    Right node -> DAG.putNode node
```

Comment:
This is _totally fine as is_, but have you considered breaking out the `Node` into a named variable for legibility?

```haskell
walkDir ignored path = do
  files <- listDirectory path

  let
    toAdd   = removeIgnored ignored files
    reducer = foldResults path ignored
    seed    = Right <| Node 
                { dataBlock = "CAE="
                , links     = [] 
                }

  foldM reducer seed toAdd >>= \case
    Left  err  -> return <| Left err
    Right node -> DAG.putNode node
```

Reply:
Love it :ok_hand: